src/NuGetGallery/Controllers/PackagesController.cs

```
@@ -1638,50 +1638,58 @@ private void SendAddPackageOwnerNotification(PackageRegistration package, User n
 
-            // delete the uploaded binary in the Uploads container
-            await _uploadFileService.DeleteUploadFileAsync(currentUser.Key);
+                    if (!(_config.AsynchronousPackageValidationEnabled && _config.BlockingAsynchronousPackageValidationEnabled))
+                    {
+                        // notify user unless async validation in blocking mode is used
+                        _messageService.SendPackageAddedNotice(package,
+                            Url.Package(package.PackageRegistration.Id, package.NormalizedVersion, relativeUrl: false),
+                            Url.ReportPackage(package.PackageRegistration.Id, package.NormalizedVersion, relativeUrl: false),
+                            Url.AccountSettings(relativeUrl: false));
+                    }
 
-            _telemetryService.TrackPackagePushEvent(package, currentUser, User.Identity);
+                    // delete the uploaded binary in the Uploads container
+                    await _uploadFileService.DeleteUploadFileAsync(currentUser.Key);
 
-            TempData["Message"] = String.Format(
-                CultureInfo.CurrentCulture, Strings.SuccessfullyUploadedPackage, package.PackageRegistration.Id, package.Version);
+                    _telemetryService.TrackPackagePushEvent(package, currentUser, User.Identity);
 
-            return Json(new
-            {
-                location = Url.Package(package.PackageRegistration.Id, package.NormalizedVersion)
-            });
+                    TempData["Message"] = String.Format(
+                        CultureInfo.CurrentCulture, Strings.SuccessfullyUploadedPackage, package.PackageRegistration.Id, package.Version);
+
+                    return Json(new
+                    {
+                        location = Url.Package(package.PackageRegistration.Id, package.NormalizedVersion)
+                    });
+                }
+                catch (Exception e)
```

is it possible to catch any specific exceptions here before general exception, so that we can give customized message?

---

Also, maybe we should propagate the error message out of `UserSafeException`?

---

It's definitely possible. We just need to enumerate which cases we want specific error messages for.